Mathlib/RingTheory/Polynomial/Pochhammer.lean

```diff
@@ -372,4 +372,49 @@ theorem descPochhammer_int_eq_ascFactorial (a b : ℕ) :
   rw [← Nat.cast_add, descPochhammer_eval_eq_descFactorial ℤ (a + b) b,
     Nat.add_descFactorial_eq_ascFactorial]
 
+/- If the Pochhammer function is evaluated at a sufficiently large non-postive integer, then it is
+zero. -/
+theorem ascPochhammer_eq_zero_of_nonpos_int {R : Type u} [Ring R] {n : ℕ} {k : R}
+    (hnk : ∃ kn : ℤ, kn ≤ 0 ∧ k = kn ∧ n ≥ 1 - kn) : (ascPochhammer R n).eval k = 0 := by
```

Comment:
I would be inclined to rewrite this as
```suggestion
    (hnk : ∃ kn : ℤ, 1 - n ≤ k ∧ kn ≤ 0 ∧ k = kn) : (ascPochhammer R n).eval k = 0 := by
```

Reply:
Can I get a sense of why you prefer it? I understand rewriting the inequality, but to me, limiting kn to a non-positive integer makes sense directly after the existential binder of kn.

Reply:
Of course I'm not the author of the suggestion but one point is that except for certain very specific situations, we avoid using `>` and `≥` and instead use `<` and `≤`. This reduces the need for rewrites.

(Aside: speaking of rewrites, I notice that elsewhere you're using the long-form `rewrite` tactic: we always use `rw` instead.)

Another point I notice here is that it's usually unhelpful to pass a raw existential as hypothesis: instead just demand the data that demonstrates it is true. Finally (and this point is, I think, subjective) I would probably just use a natural number rather than non-positive integer. With all these in mind, I'd state this lemma as:
```lean
theorem ascPochhammer_eval_neg_coe_nat_of_le {R : Type u} [Ring R] {n : ℕ} {k : ℕ} (h : 1 + k ≤ n) :
    (ascPochhammer R n).eval (-(k : R)) = 0 := by
```
or if you aren't convinced by the subjective point maybe:
```lean
theorem ascPochhammer_eval_coe_int_of_le_of_le {R : Type u} [Ring R] {n : ℕ} {k : ℤ}
    (h₁ : k ≤ 0) (h₂ : 1 - k ≤ n) : (ascPochhammer R n).eval (k : R) = 0 :=
```

As a final remark, I'd like to say I'm delighted and very grateful to see this PR! I have a backlog of work and review at the moment so I will have to wait till the weekend to give thorough review.

Reply:
Thanks for the feedback, I'll replace `rewrite` with `rw`, flip the inequalities and expand the existential as you say. I'm slightly unconvinced by the use of a natural in the way you do, what do you think of this?
```
theorem ascPochhammer_eval_neg_coe_nat_of_lt {R : Type u} [Ring R] {n k : ℕ} {r : R} (hr : r = -↑k) 
    (h : k < n) :
    (ascPochhammer R n).eval r = 0 := by sorry
```
The problem with the goal having `-(k : R)` in the result is that it becomes awkward to manipulate expressions that aren't in that exact form (and you can't use `rw`), whereas taking it into a hypothesis allows delaying it by easily moving it into a goal. I'm sure this situation must come up a lot in Mathlib, so I don't know what the best practice is, but it feels clunky to work with otherwise.

Reply:
The hypothesis `hr` is of the form `x = f y` for some function `f`. To use a such a lemma you must have (or be able to construct) the equation `x = f y` and so you should just rewrite your goal before calling the lemma. Otherwise we have a less general lemma and potentially need to write lemmas for many functions `f`.

So I still think the form I proposed is the way to go.

Reply:
Alright, this will need a bit of refactoring then 😁